test/utils/TestUniversalResolver.js

```diff
@@ -0,0 +1,87 @@
+const ENS = artifacts.require('./registry/ENSRegistry.sol');
+const PublicResolver = artifacts.require('PublicResolver.sol');
+const NameWrapper = artifacts.require('DummyNameWrapper.sol');
+const UniversalResolver = artifacts.require('UniversalResolver.sol');
+const DummyOffchainResolver = artifacts.require('DummyOffchainResolver.sol');
+
+const namehash = require('eth-ens-namehash');
+const sha3 = require('web3-utils').sha3;
+const ethers = require('ethers');
+const { dns } = require("../test-utils");
+
+contract('UniversalResolver', function (accounts) {
+
+    let ens, publicResolver, universalResolver, dummyOffchainResolver, nameWrapper;
+
+    beforeEach(async () => {
+        node = namehash.hash('eth');
+        ens = await ENS.new();
+        nameWrapper = await NameWrapper.new();
+        publicResolver = await PublicResolver.new(ens.address, nameWrapper.address);
+        universalResolver = await UniversalResolver.new(ens.address);
+        dummyOffchainResolver = await DummyOffchainResolver.new();
+
+        await ens.setSubnodeOwner('0x0', sha3('eth'), accounts[0], {from: accounts[0]});
+        await ens.setSubnodeOwner(namehash.hash('eth'), sha3('test'), accounts[0], {from: accounts[0]});
+        await ens.setResolver(namehash.hash('test.eth'), publicResolver.address, {from: accounts[0]});
+        await ens.setSubnodeOwner(namehash.hash('test.eth'), sha3('sub'), accounts[0], {from: accounts[0]});
+        await ens.setResolver(namehash.hash('sub.test.eth'), accounts[1], {from: accounts[0]});
+        await publicResolver.methods["setAddr(bytes32,address)"](namehash.hash('test.eth'), accounts[1], {from: accounts[0]});
+        await ens.setSubnodeOwner(namehash.hash('test.eth'), sha3('offchain'), accounts[0], {from: accounts[0]});
+        await ens.setResolver(namehash.hash('offchain.test.eth'), dummyOffchainResolver.address, {from: accounts[0]})
+    });
+
+    describe('findResolver()', () => {
+        it('should find an exact match resolver', async () => {
+            const result = await universalResolver.findResolver(dns.hexEncodeName('test.eth'));
+            expect(result).to.equal(publicResolver.address);
+        });
+
+        it('should find a resolver on a parent name', async () => {
+            const result = await universalResolver.findResolver(dns.hexEncodeName('foo.test.eth'));
+            expect(result).to.equal(publicResolver.address);
+        });
+
+        it('should choose the resolver closest to the leaf', async () => {
+            const result = await universalResolver.findResolver(dns.hexEncodeName('sub.test.eth'));
```

Comment:
Probably easier to understand if the test name is 'foo.sub.test.eth' so that it's taking from the closer parent(sub.test.eth not from test.eth) rather than the resolver set to the domain itself